### PR TITLE
Normalize BlockPolicyNet by upper and lower bounds

### PR DESCRIPTION
**Pull request by sbenthall**
This PR addresses #66 

Now, when a BlockPolicyNet is created for a block with a control that has an upper bound, lower bound, or both, the network will output a value within the bounds, and the gradients will work out properly so the policy can be trained into the binding constraints.

There are a few different contributions here:
- Test cases for bounded blocks in `conftest.py`
- Adding an optional `until` argument to `block.transition()` which let you simulate forward the block dynamics until hitting a particular symbol. This is important because to get at the values in the information set of the control, it's important to not overwrite the post-states, which are by definition downstream of the control.
- A new utility `utils.create_vectorized_function_wrapper_with_mapping` -- not elegantly named -- which wraps a function (here, the upper and lower bound functions) and a mapping from symbols to tensor indices, which allows the function to be executed properly in the 'scope' of the tensor. This is needed in the forward function of the neural network, where the normalization of the output happens given the input vector, as opposed to a `dict` representation of the inputs.
- The changes to `BlockPolicyNet` to implement the normalization. A lot of work happens on initialization. This also means overriding the default `forward` function of the `Net` class. This makes a cross-class assumption of network architecture, which is going to complicate #52 but so it goes.

### CODE REVIEW

**Comment by sbenthall** [comment]
The other point that I think would improve things: if the `forward()` method of the `BlockPolicyNet` refers to `super.forward()`, then it won't have to overwrite the superclass `forward()` method or depend on a specific ANN architecture, making #52 easier.

File: tests/test_ann.py [hunk 1]
```diff
@@ -174,6 +171,104 @@ def test_case_3_2(self):
 
         self.assertTrue(torch.allclose(c_ann.flatten(), given_m.flatten(), atol=0.04))
 
+    def test_case_5_double_bounded_upper_binds(self):
```
[the remark this answers is not in the document]
Comment (sbenthall) [reply tests/test_ann.py:174]: https://github.com/scikit-agent/scikit-agent/issues/79

**Comment by sbenthall** [comment]
Fixes based on the first round of reviews are in.

**Comment by alanlujan91** [comment]
Inclusive vs exclusive constraints, if the constraint binds optimally, can this method actually land at the constraint? or just very close to the constraint?

**Comment by sbenthall** [comment]
There seem to be several ways of dealing with constraints.

- the sigmoid and softplus functions exclude the constraints (0, 1) but have the benefit of non-zero gradients in their entire domain. The network will get arbitrarily close to the bounds. Since neural networks are numerical approximators, this is not so bad.
- We could replace these with tranformation functions that are inclusive of the bounds ([hardtanh](https://docs.pytorch.org/docs/stable/generated/torch.nn.Hardtanh.html), rectified linear)). These have the downside that for large areas (i.e., when the function hits the bound) the gradient is 0, which can prevent further training. There are some workarounds to this, which involve more complex transformations that act differently on the forward and backwards passes. But this is apparently a classic problem in deep learning; this might just be a bullet we have to bite for deep learning solvers.
- There are ways to work the constraints into the _loss function_, but I see that as not as user-friendly as something that just works given the model definition. I'm looking at the Fisher-Burmeister function you've pointed out

image

This works for single-bounded variables. It adds a lot of complexity to the loss function and leaves the space of decisions technically unbounded. I'm skeptical of this method, because in the AiO loss function, this loss will be balanced with the other losses (due to the value function residuals, etc.) and there's nothing to guarantee that the decision rule created a by  training regime that ends at some $\epsilon$ tolerance level will in fact be within the constraints. Unless I'm missing something.

In sum, there are more than one way to address (or not address) the issue of inclusive bounds, but none of them are perfect, as this is a general problem of deep learning. 

If the current implementation is _documented_ so that it's clear that it considers the bounds to be exclusive, would this suffice? A separate issue could be created to handle the inclusive bound cases.

(Perhaps @alanlujan91 you've already worked the FB function into your AiO loss functions; in that case, would you want a manual override of default bounding on the PolicyNet? That could be done without too much difficulty.)
